File: hbTapi Components/Delphi Samples/Professional Edition/Location/uMain.pas
{-------------------------------------------------------------------------
   Copyright by Haeger + Busch, Germany                / >>>>>>>>> /-----
   Ingenieurbuero fuer Kommunikationslösungen         / <<<<<<<<< /
 ----------------------------------------------------/ >>>>>>>>> /
   Homepage  : http://www.hbTapi.com
   EMail     : info@hbTapi.com
   Package   : hbTapi Components
 -------------------------------------------------------------------------}
 
unit uMain;

interface

uses
  Windows, Messages, SysUtils, Classes, Graphics, Controls, Forms, Dialogs, hbtapi,
  StdCtrls, ExtCtrls;

type
  TForm1 = class(TForm)
    GroupBox1: TGroupBox;
    Label_LocationName: TLabel;
    Button1: TButton;
    Button2: TButton;
    GroupBox2: TGroupBox;
    Label3: TLabel;
    Button3: TButton;
    ListBox1: TListBox;
    Label5: TLabel;
    Label6: TLabel;
    Label7: TLabel;
    Label8: TLabel;
    Label9: TLabel;
    Label10: TLabel;
    Label_PermanentLocationID: TLabel;
    Label_CountryID: TLabel;
    Label_CountryCode: TLabel;
    Label_CountryName: TLabel;
    Label_LocalAccessCode: TLabel;
    Label_LongDistanceAccessCode: TLabel;
    Label2: TLabel;
    Label_CityCode: TLabel;
    Button4: TButton;
    hbTapiLine1: ThbTapiLine;
    GroupBox3: TGroupBox;
    Edit1: TEdit;
    Button5: TButton;
    procedure Button1Click(Sender: TObject);
    procedure FormCreate(Sender: TObject);
    procedure FormDestroy(Sender: TObject);
    procedure DoUpdateSystemLocation(Sender: TObject);
    procedure DoUpdateLocations(Sender: TObject);
    procedure Button4Click(Sender: TObject);
    procedure Button5Click(Sender: TObject);
    procedure hbTapiLine1TapiTranslateChange(Sender: TObject);
  private
    FLocationList   : ThbTapiLocationList;
    procedure UpdateSystemLocation;
    procedure UpdateLocations;
  public
    { Public-Deklarationen }
  end;

var
  Form1: TForm1;

implementation

{$R *.DFM}

procedure TForm1.FormCreate(Sender: TObject);
begin
  FLocationList   := ThbTapiLocationList.Create;
//  hbTapiLine1.DeviceID := 0;  // we have to select a device to change the location
  hbTapiLine1.Enabled := True;
  UpdateSystemLocation;
  UpdateLocations;
end;

procedure TForm1.FormDestroy(Sender: TObject);
begin
  FLocationList.Free;
end;

procedure TForm1.UpdateSystemLocation;
begin
  with hbTapiLine1.Location do
  begin
    Label_LocationName.Caption            := LocationName;
    Label_PermanentLocationID.Caption     := IntToStr(LocationID);
    Label_CountryID.Caption               := IntToStr(CountryID);
    Label_CountryName.Caption             := CountryName;
    Label_CountryCode.Caption             := CountryCode;
    Label_LocationName.Caption            := LocationName;
    Label_CityCode.Caption                := CityCode;
    Label_LocalAccessCode.Caption         := LocalAccessCode;
    Label_LongDistanceAccessCode.Caption  := LongDistanceAccessCode;
  end;
end;

procedure TForm1.UpdateLocations;
var i: integer; s: String;
begin
  ListBox1.Items.Clear;
  with FLocationList do
  try
    Update;
    for i := 0 to Count-1 do
    begin
      s := Items[i].LocationName + ', ID=' + IntToStr(Items[i].LocationID);
      if Items[i].LocationID = CurrentLocationID then
        s := s + ', CurrentLocation';
      ListBox1.Items.AddObject(s, TObject(Items[i].LocationID));
    end;
  except
  end;
end;

procedure TForm1.Button1Click(Sender: TObject);
begin
  with hbTapiLine1 do
  begin
    TranslateDialog(Edit1.Text, Handle);
    UpdateSystemLocation;
    UpdateLocations;
  end;
end;

procedure TForm1.DoUpdateSystemLocation(Sender: TObject);
begin
  try
    hbTapiLine1.Location.Update;
    UpdateSystemLocation;
  except
  end;  
end;

procedure TForm1.DoUpdateLocations(Sender: TObject);
begin
  UpdateLocations;
end;

procedure TForm1.Button4Click(Sender: TObject);
var i: integer;
begin
  i := ListBox1.ItemIndex;
  if i >= 0 then
    hbTapiLine1.Location.LocationID := DWord(ListBox1.Items.Objects[i]);
  UpdateSystemLocation;
  UpdateLocations;
end;

procedure TForm1.Button5Click(Sender: TObject);
begin
  MessageDlg('Dialable address: ' + hbTapiLine1.TranslateAddress(Edit1.Text), mtInformation, [mbOK], 0);
end;

procedure TForm1.hbTapiLine1TapiTranslateChange(Sender: TObject);
begin
  MessageDlg('TAPI Translate Caps changed', mtInformation, [mbOK], 0);
  UpdateSystemLocation;
  UpdateLocations;
end;

end.

File: hbTapi Components/BCB Samples/Professional Edition/Location/uMain.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "uMain.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "hbTapi"
#pragma resource "*.dfm"
TForm1 *Form1;
//---------------------------------------------------------------------------
__fastcall TForm1::TForm1(TComponent* Owner)
        : TForm(Owner)
{
}

void __fastcall TForm1::FormCreate(TObject *Sender)
{
  FLocationList = new ThbTapiLocationList();
  hbTapiLine1->DeviceID = 0;  // we have to select a device to change the location
  UpdateSystemLocation();
  UpdateLocations();

}
//---------------------------------------------------------------------------

void __fastcall TForm1::UpdateLocations()
{
  int i; String s;
  ListBox1->Items->Clear();
  FLocationList->Update();
  for (i=0;i<FLocationList->Count;i++)
  {
    s = FLocationList->Items[i]->LocationName + ", ID=" + IntToStr((int)FLocationList->Items[i]->LocationID);
    if (FLocationList->Items[i]->LocationID == FLocationList->CurrentLocationID)
      s = s + ", CurrentLocation";
    ListBox1->Items->AddObject(s, (TObject*)FLocationList->Items[i]->LocationID);
  }
}

void __fastcall TForm1::UpdateSystemLocation()
{
    Label_LocationName->Caption            = hbTapiLine1->Location->LocationName;
	Label_PermanentLocationID->Caption     = IntToStr((int)hbTapiLine1->Location->LocationID);
    Label_CountryID->Caption               = IntToStr((int)hbTapiLine1->Location->CountryID);
    Label_CountryName->Caption             = hbTapiLine1->Location->CountryName;
    Label_CountryCode->Caption             = hbTapiLine1->Location->CountryCode;
    Label_LocationName->Caption            = hbTapiLine1->Location->LocationName;
    Label_CityCode->Caption                = hbTapiLine1->Location->CityCode;
    Label_LocalAccessCode->Caption         = hbTapiLine1->Location->LocalAccessCode;
    Label_LongDistanceAccessCode->Caption  = hbTapiLine1->Location->LongDistanceAccessCode;
}

//---------------------------------------------------------------------------
void __fastcall TForm1::Button1Click(TObject *Sender)
{
  hbTapiLine1->TranslateDialog(Edit1->Text, Handle);
  UpdateSystemLocation();
  UpdateLocations();
}
//---------------------------------------------------------------------------

void __fastcall TForm1::FormDestroy(TObject *Sender)
{
  delete FLocationList;

}
//---------------------------------------------------------------------------

void __fastcall TForm1::Button2Click(TObject *Sender)
{
  try
  {
    hbTapiLine1->Location->Update();
    UpdateSystemLocation();
  }
  catch (...)
  {
  }
}
//---------------------------------------------------------------------------

void __fastcall TForm1::Button4Click(TObject *Sender)
{
  int i;
  i = ListBox1->ItemIndex;
  if (i >= 0)
    hbTapiLine1->Location->LocationID = (unsigned) ListBox1->Items->Objects[i];
  UpdateSystemLocation();
  UpdateLocations();

}
//---------------------------------------------------------------------------

void __fastcall TForm1::Button5Click(TObject *Sender)
{
  MessageDlg("Dialable address: " + hbTapiLine1->TranslateAddress(Edit1->Text, 0), mtInformation, TMsgDlgButtons() << mbOK, 0);

}
//---------------------------------------------------------------------------

void __fastcall TForm1::Button3Click(TObject *Sender)
{
  UpdateLocations();

}
//---------------------------------------------------------------------------

